pilot/pkg/model/gateway.go

```diff
@@ -45,24 +56,101 @@ func MergeGateways(gateways ...Config) *MergedGateway {
+			if gatewayPorts[s.Port.Number] {
+				// We have two servers on the same port. Should we merge?
+				// 1. Yes if both servers are plain text and HTTP
+				// 2. Yes if both servers are using TLS
+				//    if using HTTPS ensure that port name is distinct so that we can setup separate RDS
+				//    for each server (as each server ends up as a separate http connection manager due to filter chain match
+				// 3. No for everything else.
+
+				if p, exists := plaintextServers[s.Port.Number]; exists {
+					currentProto := ParseProtocol(p[0].Port.Protocol)
+					if currentProto != protocol || !protocol.IsHTTP() {
+						log.Debugf("skipping server on gateway %s port %s.%d.%s: conflict with existing server %s.%d.%s",
+							spec.Name, s.Port.Name, s.Port.Number, s.Port.Protocol, p[0].Port.Name, p[0].Port.Number, p[0].Port.Protocol)
+						continue
+					}
+					rdsName := getRDSRouteName(s)
+					rdsRouteConfigNames[rdsName] = append(rdsRouteConfigNames[rdsName], s)
+				} else {
+					// This has to be a TLS server
+					if isHTTPServer(s) {
+						rdsName := getRDSRouteName(s)
+						// both servers are HTTPS servers. Make sure the port names are different so that RDS can pick out individual servers
+						// WE cannot have two servers with same port name because we need the port name to distinguish one HTTPS server from another
+						// WE cannot merge two HTTPS servers even if their TLS settings have same path to the keys, because we don't know if the contents
+						// of the keys are same. So we treat them as effectively different TLS settings.
+						if _, exists := rdsRouteConfigNames[rdsName]; exists {
+							log.Debugf("skipping server on gateway %s port %s.%d.%s: non unique port name for HTTPS port",
+								spec.Name, s.Port.Name, s.Port.Number, s.Port.Protocol)
+							continue
+						}
+						rdsRouteConfigNames[rdsName] = []*networking.Server{s}
+					}
+					tlsServers[s.Port.Number] = append(tlsServers[s.Port.Number], s)
 				}
-				servers[s.Port.Number] = append(ss, s)
 			} else {
-				servers[s.Port.Number] = []*networking.Server{s}
+				gatewayPorts[s.Port.Number] = true
+				if isTLSServer(s) {
+					tlsServers[s.Port.Number] = []*networking.Server{s}
+				} else {
+					plaintextServers[s.Port.Number] = []*networking.Server{s}
+				}
+
+				if isHTTPServer(s) {
+					rdsRouteConfigNames[getRDSRouteName(s)] = []*networking.Server{s}
+				}
 			}
 			log.Debugf("MergeGateways: gateway %q merged server %v", name, s.Hosts)
 		}
 	}
 
+	// Concatenate both sets of servers
+	for p, v := range tlsServers {
+		servers[p] = v
+	}
+	for p, v := range plaintextServers {
+		servers[p] = v
+	}
+
 	return &MergedGateway{
-		Names:   names,
-		Servers: servers,
+		Names:               names,
+		Servers:             servers,
+		RDSRouteConfigNames: rdsRouteConfigNames,
+	}
+}
+
+func isTLSServer(server *networking.Server) bool {
+	if server.Tls == nil || ParseProtocol(server.Port.Protocol).IsHTTP() {
```

Comment:
```go
func isTLSServer(server *networking.Server) bool {
    return server.Tls != nil && !ParseProtocol(server.Port.Protocol).IsHTTP()
}
```

(which is just `!(server.Tls == nil || ParseProtocol(server.Port.Protocol).IsHTTP())`)